drv/sprot-api/src/lib.rs

```diff
@@ -333,232 +422,238 @@ impl From<u8> for MsgType {
-        Protocol::Busy => return Err(MsgError::RotBusy),
-        Protocol::V1 => {}
-        Protocol::Unsupported => return Err(MsgError::UnsupportedProtocol),
+
+    pub fn len(&self) -> usize {
+        self.buf.len()
     }
-    if buffer.len() < HEADER_SIZE + CRC_SIZE {
-        return Err(MsgError::BadMessageLength);
+
+    pub fn as_slice(&self) -> &[u8] {
+        &self.buf[..]
     }
-    if let Ok((header, payload_start)) =
-        hubpack::deserialize::<MsgHeader>(buffer)
-    {
-        let len = header.len as usize;
-        if_chain! {
-            if let Some(crc_buf) = payload_start.get(len..len + CRC_SIZE);
-            if let Ok((crc, _)) = hubpack::deserialize::<u16>(crc_buf);
-            if let Some(msg_bytes) = buffer.get(0..HEADER_SIZE+len);
-            if crc == CRC16.checksum(msg_bytes);
-            then {
-                Ok((header.msgtype, &payload_start[..len]))
-            } else {
-                Err(MsgError::InvalidCrc)
-            }
-        }
-    } else {
-        // Content didn't matter for hubpack::deserialize.
-        // So, there weren't enough bytes to decode a header.
-        Err(MsgError::BadMessageLength)
+
+    pub fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.buf[..]
     }
-}
 
-/// Parse the header from an incomplete received message
-pub fn rx_payload_remaining_mut(
-    valid_bytes: usize,
-    buffer: &mut [u8],
-) -> Result<&mut [u8], MsgError> {
-    if valid_bytes.min(buffer.len()) < HEADER_SIZE {
-        return Err(MsgError::Incomplete);
+    pub fn payload(&self, rxmsg: &VerifiedRxMsg) -> &[u8] {
+        &self.buf[HEADER_SIZE..][..rxmsg.0.payload_len as usize]
     }
-    match Protocol::from(buffer[0]) {
-        Protocol::Ignore => return Err(MsgError::NoMessage),
-        Protocol::Busy => return Err(MsgError::RotBusy),
-        Protocol::V1 => (),
-        _ => return Err(MsgError::UnsupportedProtocol),
+
+    pub fn parse_header(
+        &self,
+        valid_bytes: usize,
+    ) -> Result<MsgHeader, SprotError> {
+        // We want to be able to return `RotBusy` before `Incomplete`.
+        self.parse_protocol()?;
+        if valid_bytes < HEADER_SIZE {
+            return Err(SprotError::Incomplete);
+        }
+        let (header, _) = hubpack::deserialize::<MsgHeader>(&self.buf[..])?;
+        if header.payload_len as usize > PAYLOAD_SIZE_MAX {
+            return Err(SprotError::BadMessageLength);
+        }
+        Ok(header)
     }
-    if let Ok((header, _payload_start)) =
-        hubpack::deserialize::<MsgHeader>(buffer)
-    {
-        let end = MIN_MSG_SIZE + (header.len as usize);
-        if end > buffer.len() {
-            Err(MsgError::BadMessageLength)
-        } else if valid_bytes < end {
-            Ok(&mut buffer[valid_bytes..end])
+
+    pub fn validate_crc(&self, header: &MsgHeader) -> Result<(), SprotError> {
+        // The only way to get a `MsgHeader` is to call parse_header, which
+        // already validated the payload size.
```

Comment:
s/validated/sanity checked/

We don't know if that length is correct, only that it fits.
The CRC will validate the length (and the rest of the message).

Reply:
Fixed in [f86812d](https://github.com/oxidecomputer/hubris/pull/965/commits/f86812d56e41435275dab8b4cf32982f88d4c35f)